examples/end-to-end-examples/support_chatbot/chatbot.py

```diff
@@ -0,0 +1,546 @@
+import os
+import json
+import re
+import string
+import time
+from tqdm import tqdm
+
+import langchain
+from langchain.document_loaders import UnstructuredFileLoader
+from langchain.text_splitter import RecursiveCharacterTextSplitter
+import pickle
+from langchain.prompts import PromptTemplate
+from langchain.vectorstores import FAISS
+from langchain.chains import LLMChain, RetrievalQA
+from langchain.chains.combine_documents.stuff import StuffDocumentsChain
+from langchain.schema import Document, BaseRetriever
+import sys
+
+ROOT_DIR = os.path.dirname(os.path.abspath(__file__))
+sys.path.append(ROOT_DIR) 
+from repo_downloader import RepoDownloader 
+from web_downloader import WebScraper
+from oci_converser import OCIObjectStorageManager
+
+class RetrieverWithScore(BaseRetriever):
+    def __init__(self, 
+                 search_type: str,
+                 vector_store: FAISS,
+                 k: int,
+                 score_threshold: int):
+        self.k = k
+        self.vector_store=vector_store
+        self.score_threshold = score_threshold
+        self.search_type=search_type
+
+    def _get_relevant_documents(self, query: str) -> list[Document]:
+        # [NOTE] we removed the search type, only use search_type = "similarity"
+        if self.search_type != "similarity":
+            raise ValueError(f"Only search_type='similarity' is supported with scores")
+        docs_and_scores = self.vector_store.similarity_search_with_score(query=query, 
+                                                                         k=self.k, 
+                                                                         score_threshold=self.score_threshold)
+        for doc, score in docs_and_scores:
+            doc.metadata = {**doc.metadata, **{"score": score}}
+        return [doc for (doc, _) in docs_and_scores]
+
+    def aget_relevant_documents(self, query):
```

Comment:
whats this function for?

Reply:
Mandatory from LangChain, though not used I believe. Python thing ~~
TypeError: Can't instantiate abstract class RetrieverWithScore with abstract method aget_relevant_documents